Sources/SpeziDevices/HealthMeasurements.swift

```diff
@@ -0,0 +1,344 @@
+
+            guard let bloodPressureSample else {
+                logger.debug("Discarding invalid blood pressure measurement ...")
+                return nil
+            }
+
+            logger.debug("Measurement loaded: \(String(describing: measurement))")
+
+            healthKitMeasurement = .bloodPressure(bloodPressureSample, heartRate: heartRateSample)
+        }
+
+        // prepend to pending measurements
+        pendingMeasurements.insert(healthKitMeasurement, at: 0)
+        return healthKitMeasurement.id
+    }
+
+    /// Discard a pending measurement.
+    ///
+    /// Measurements are discarded if they are no longer of interest. Either because the users discarded the measurements contents or
+    /// if the measurement was processed otherwise (e.g., saved to an external data store).
+
+    /// - Parameter measurement: The pending measurement to discard.
+    /// - Returns: Returns `true` if the measurement was in the array of pending measurement, `false` if nothing was discarded.
+    @MainActor
+    @discardableResult
+    public func discardMeasurement(_ measurement: HealthKitMeasurement) -> Bool {
+        guard let index = self.pendingMeasurements.firstIndex(where: { $0.id == measurement.id }) else {
+            return false
+        }
+        let element = self.pendingMeasurements.remove(at: index)
+
+        let id = element.id // we need to capture id, element.id results in #Predicate to not compile
+        do {
+            try modelContainer?.mainContext.delete(
+                model: StoredMeasurement.self,
+                where: #Predicate<StoredMeasurement> { $0.associatedMeasurement == id }
+            )
+        } catch {
+            logger.error("Failed to remove measurement from storage: \(error)")
+        }
+
+        return true
+    }
+}
+
+
+extension HealthMeasurements: Module, EnvironmentAccessible, DefaultInitializable {}
+
+
+extension HealthMeasurements {
+    @MainActor
+    func refreshFetchingMeasurements() throws {
+        pendingMeasurements.removeAll()
+        if let modelContainer, modelContainer.mainContext.hasChanges {
+            try modelContainer.mainContext.save()
+        }
+        fetchMeasurements()
+    }
+
+    @MainActor
+    private func fetchMeasurements() {
+        guard let modelContainer else {
+            return
+        }
+
+        var fetchAll = FetchDescriptor<StoredMeasurement>()
+        // TODO: fetchAll.includePendingChanges = true
+
+        let context = modelContainer.mainContext
+        let storedMeasurements: [StoredMeasurement]
+        do {
+            storedMeasurements = try context.fetch(fetchAll)
+        } catch {
+            logger.error("Failed to retrieve stored measurements from disk \(error)")
+            return
+        }
+
+        print("hasChanges1: \(context.hasChanges == true)")
+        try? context.save()
+
+        for storedMeasurement in storedMeasurements {
+            print("Looking at \(storedMeasurement)")
```

Comment:
Looks like fun with Swift Data debugging. As noted in Slack; we might want to separate the measurement storage in a separate module and/or add this in a separate PR if it might get challenging to include it in this PR within a reasonable amount of time.